EIPS/eip-7512.md

```diff
@@ -18,71 +18,89 @@ The proposal aims to create a standard for an onchain representation of audit re
+    - `authors`: A list of authors that contributed to this audit. This SHOULD be the persons who audited the contracts and created the audit
 - Audit
-    - `auditor` - Information on the auditor
-    - `contract` - MUST be the `chainId` and `address` of the deployed contract the audit is related to.
-    - `issuedAt` - MUST contain the information when the original audit (identified by the `auditHash`) was issued
-    - `ercs` - A list of ERCs that are implemented by the target contract. The ERCs listed MUST be fully implemented. This list MAY be empty.
-    - `auditHash` - MUST be the hash of the original audit. This allows to verify that this onchain information belongs to a specific audit.
-    - `auditUri` - SHOULD point to a source where the audit can be retrieved
+    - `auditor`: Information on the auditor
+    - `auditedContract`: MUST be the `chainId` as well as `implementation` of the deployed contract the audit is related to
+    - `issuedAt`: MUST contain the information when the original audit (identified by the `auditHash`) was issued
+    - `ercs`: A list of ERCs that are implemented by the target contract. The ERCs listed MUST be fully implemented. This list MAY be empty
+    - `auditHash`: MUST be the hash of the original audit. This allows onchain verification of information that may belong to a specific audit
+    - `auditUri`: SHOULD point to a source where the audit can be retrieved
+- Contract
+    - `chainId`: MUST be a `bytes32` representation of the [ERC-155](./eip-155.md) chain ID of the blockchain that the contract has been deployed in
+    - `implementation`: MUST be a `bytes32` representation of a contract's implementation as described below
+
+### Contract Implementation
+
+Security audits generally cover a specific snapshot of an overall codebase, meaning that consequent updates of the code are not to be considered covered by it. While smart contracts are immutable on a blockchain, multiple EIP standards have been defined (i.e. ERC, ERC, etc.) that permit a "circumvention" of this blockchain trait and allow the logic of a smart contract to be updated. 
+
+In this regard, simply associating a security audit with an `address` of a smart contract is insufficient as it may ultimately change its underlying implementation. In order to future-proof this EIP and cover certain cases such as polymorphic contracts that permit a contract to be re-deployed in the same `address`, we provide the following definition of a contract `implementation`.
+
+An `implementation` is a `bytes32` representation of a `keccak256` hash-chain of the contract's `address` and bytecode hashes derived using the `EXTCODEHASH` opcode ([ERC-1052](./eip-1052.md)). To elaborate, each "contract" can be thought of as consisting of an `address` that it is deployed in as well as its bytecode. When dealing with proxies, diamond patterns, and other such approaches we end up with "contracts" that ultimately relay calls to other "contracts".
```

Comment:
The initial version of this ERC actually used `runtimeCode` instead of addresses (see https://hackmd.io/@rimeissner/erc-onchain-audit-representation). Here the comment was that due to the contract state the `runtimeCode` cannot be considered sufficient, and that some teams do checks on the deployment setup for a specific address.

It seems different auditors have different processes here and the question is how to include this in an initial version. 

Potentially either having both, or allowing to choose will be necessary.

Personally I would opt for a combination of `chainid`, `address`, `runtimeCode` (and optionally even `deploymentCode`, but deployment code might be meaningless if a unknown factory was used)

Reply:
As this standard is meant to represent a ubiquitous case, I believe we should stick with solely the deployed code. Not all audit companies validate the actual arguments of the deployed contracts and the standard specifying that they are required would exclude them which can be considered unfair.

The current proposed case associates the combination you specified (`chainId`, `address`, `runtimeCode`) by having the `chainId` as a dedicated entry and the combination of an `address` and `runtimeCode` as part of the hash payload. A hash of the `runtimeCode` (which `EXTCODEHASH` yields) is equivalent to it and much cheaper than loading the full code in Solidity implementations.

Reply:
ohh yeah I meant the `runtimeCodeHash` ... typo from my side :see_no_evil: 

Just to clarify, so you would argue not to add `deploymentCode` (or anything similar). The combination of `chainId`, `address`, `runtimeCodeHash` would work. 

Should `address` be optional?

Reply:
Yes, the `deploymentCode` should be avoided as it cannot be extracted on-chain and also requires additional effort from audit companies that may not be willing to provide it. The audit summary or another data point of the audit itself could have a `bool` flag that indicates the actual arguments of the deployment were "audited" or not.

RE: `address`, as this is a hash-chain, rendering it optional would require all integrators to calculate two hashes instead of one whenever evaluating an unknown contract (as an auditor may have submitted it with the `address` associated and without).

Additionally, we will face issues where the same bytecode has been audited multiple times by different auditors if we solely rely on the `runetimeCodeHash`. 

Based on the above arguments, I think the `address` should be mandatory.